lib/memory.cpp

```diff
@@ -234,6 +236,13 @@ void Memory::store(Word addr, uint8_t data_type, Word data) {
 }
 
 void Memory::update_peripherals() {
+    //Only actually update/check peripherals every 65535 times this function is called
+    //This is since, although non-blocking, the read system call in the UART is REALLY REALLY REALLY slow
+    //FIXME does this introduce software problems? I don't think so, no chance some types that fast anyways and
+    //regardless it likely wouldn't even matter
+    ++this->m_delay_uart_update_counter;
+    if (this->m_delay_uart_update_counter) return;//Counter didn't overflow, so don't update the timer
```

Comment:
Yeah you're not wrong about the readability, and yep, the performance gain you mentioned aligns with my intuition as well.

Even as-is, in `nouveau_stress_test` this line (and the similar one like it in `csr.cpp`) is responsible for around 2% of instructions executed. I will do some testing to see if adding a configurable constant is prohibitively expensive (but as you say it likely isn't).

I may also implement this fix (if it works out) for `csr.cpp` as well, or even combine the two so we only pay the cost once.

Reply:
Implementing the fix and using one counter for both the timer and for the UART actually _improves_ performance. Perhaps because overflow is kind of undefined behaviour, so the kernel is more confident in optimizing things this way?

Either way it has been fixed! :)